Refactors transaction step for saving admin data

afred opened the pull request:
By default, multi-valued input fields display an additional set of input fields that are blank. These get submitted by the form and are usually discarded, but because we are treating Sony Ci ID field special, it was erroneously saving additional blank values for Sony Ci ID, which was resulting in multiple players of the same media dislaying on AAPB when the record was published.

Multiple Sony Ci IDs are allowed in cases of mult-part Assets. They are stored as a serialized JSON array in the AdminData.sonyci_id field, which is related to Asset reords via a global ID field on AdminData.gid. And the global ID fields was used instead of a normal foreign key because the AssetResource used to be just Asset, and stored in Fedora, not the database, and a global ID was the closest thing to a foreign key across 2 persistence layers that we could think of :)

Fixes #882.

afred committed:
Refactors transaction step for saving admin data

By default, multi-valued input fields display an additional set of input fields that are blank.
These get submitted by the form and are usually discarded, but because we are treating Sony Ci
ID field special, it was erroneously saving additional blank values for Sony Ci ID, which was
resulting in multiple players of the same media dislaying on AAPB when the record was published.

Multiple Sony Ci IDs are allowed in cases of mult-part Assets. They are stored as a serialized JSON
array in the AdminData.sonyci_id field, which is related to Asset reords via a global ID field on
AdminData.gid. And the global ID fields was used instead of a normal foreign key because the
AssetResource used to be just Asset, and stored in Fedora, not the database, and a global ID was
the closest thing to a foreign key across 2 persistence layers that we could think of :)

Fixes #882.

afred commented:
@bkiahstroud and/or @orangewolf could I get an extra set of 👁️ on the logic here? This is some consequential code as it handles not only our Asset relations to Sony Ci media, but the Asset relation to the Bulkrax Importers and Batch Ingest Batches as well.

I think i tested all possibilities of adding/removing Sony Ci IDs, but did not do any kind of regression testing on persistence of bulkrax identifier or other admin data fields.

bkiahstroud commented on app/transactions/ams/steps/create_aapb_admin_data.rb:
```diff
@@ -51,22 +51,26 @@ def save_aapb_admin_data(change_set)
+        AdminData.attributes_for_update.each do |field|
+          field = field.to_s
+          # Convert emtpy strings to nil, whether the value is a scalar or an array
+          new_admin_data_value = if change_set.fields[field].respond_to?(:select)
```
Keying off of `respond_to?(:select)` will likely cast a much wider net than `SERIALIZED_FIELDS` does. Do we understand which fields will be impacted and how they will respond to this change?

bkiahstroud commented on app/transactions/ams/steps/create_aapb_admin_data.rb:
```diff
@@ -51,22 +51,26 @@ def save_aapb_admin_data(change_set)
+        AdminData.attributes_for_update.each do |field|
+          field = field.to_s
+          # Convert emtpy strings to nil, whether the value is a scalar or an array
+          new_admin_data_value = if change_set.fields[field].respond_to?(:select)
+            change_set.fields[field].reject {|v| v.blank? }
+          elsif change_set.fields[field].blank?
+            nil
+          end
```
Can you confirm if I'm reading this conditional correctly? 

```
if can be multiple
  filter out blank values
else if single and blank
  set new value to nil
end
```

If I am reading that correctly, I think we're missing a scenario, which would be "is single and **not** blank". In that case, I think we just lose the value. I think something like this would resolve this: 

```suggestion
          new_admin_data_value = if change_set.fields[field].respond_to?(:select)
            change_set.fields[field].reject {|v| v.blank? }
          elsif change_set.fields[field].blank?
            nil
          else # non-multiple with value present
            change_set.fields[field]
          end
```

bkiahstroud commented on app/transactions/ams/steps/create_aapb_admin_data.rb:
```diff
@@ -51,22 +51,26 @@ def save_aapb_admin_data(change_set)
+            nil
+          end
+
+          if can_empty_field?(field) || new_admin_data_value.present?
+            admin_data.write_attribute(field, new_admin_data_value)
```
Since we only write conditionally, I'd love to see an `else` which logs fields we choose _not_ to write. Historically, this has helped immensely when trying to track down issues along the lines of "I see my data going in but not coming out" 

```suggestion
          if can_empty_field?(field) || new_admin_data_value.present?
            admin_data.write_attribute(field, new_admin_data_value)
          else
            Rails.logger.debug("Chose not to update AdminData #{admin_data.id}'s field #{field} with value(s): #{change_set.fields[field]}")
```

bkiahstroud reviewed (changes requested):
I left a few questions and suggestions. Also, it looks like some relevant specs are failing in CI

afred committed:
Handles non-multiple, non-blank (default) case

Co-authored-by: Kiah Stroud <32469930+bkiahstroud@users.noreply.github.com>

afred committed:
Add debug log for when nothing gets written

Co-authored-by: Kiah Stroud <32469930+bkiahstroud@users.noreply.github.com>

afred committed:
Enumerates over SERIALIZED_FIELDS explicitly rather than duck typing

afred committed:
Another refactor/simplification

Uses AdminData.attributes_for_updates as a filter to the change set fields.
Filters out blank sony ci ids so they don't get serialized.
Adds validation routine to AdminData for guarding against fields that cannot
  be deleted once set: bulkrax_importer_id and batch_ingest_batch_id.
  These IDs are foreign keys and if they accidentally get unset, we loase
  relational integrity, so need to make sure they are not accidentally
  getting unset via ingest, or updating Assets via UI, or anywhere else.

afred commented:
@bkiahstroud even more of a simplification...

* Explicitly checks for `sonyci_id` field rather than using `AdminData::SERIALIZED_FIELDS`. This is an unnecessary abstraction that was put in place with the anticipation that we'd end up having more than one serialized field. It's been many years, and we have not, so I don't mind getting rid of that little bit of abstraction since that's the only place it's used. Furthermore, I think we'd be inclined to use regular JSON fields rather than serialized fields should an additional need arise.
* Moves the logic for "you cannot delete values for some fields after they've been set" into validation method `AdminData#validate_undeletable_fields`, which is a more appropriate location since the logic should apply no matter where the model is being used for persistence. This eliminates the need for the `should_empty_admin_data_value` method and reduces logic complexity in `set_admin_data_attributes`
* Removes unnecessary conversion of `AdminData#attributes_for_update` values from string, to symbol, back to string -- just keep them as strings, and we should be good.
* Uses `AdminData#attributes_for_update` to slice out AdminData params from the `AssetResourceForm` rather than looping over it, allowing us to update the

afred commented:
Specs are passing for me locally, so idk what's up with the failures here. Can you replicate the failures locally?

bkiahstroud commented:
> Specs are passing for me locally, so idk what's up with the failures here. Can you replicate the failures locally?

I can, I get the same failures locally

bkiahstroud commented on app/services/aapb/batch_ingest/csv_item_ingester.rb:
```diff
@@ -213,8 +213,8 @@ def set_batch_ingest_id_on_related_asset(work_id, ability)
         def set_admin_data_attributes(admin_data, attributes)
           # add existing admin_data values so they're preserved in the AssetActor
           AdminData.attributes_for_update.each do |admin_attr|
-            next unless attributes.keys.include?(admin_attr.to_s)
-            admin_data.send("#{admin_attr}=", attributes[admin_attr.to_s])
```
Why remove the `#to_s`'s here? If I put them back, one of the three failing specs pass

afred replied:
it shouldn't be necessary. I found that they start as strings, get converted to symbols, but then in all comparisons (except maybe in a spec somewhere) they are compared as strings. I'm guessing the symbols were used thinking it would be a best practice to store field names as an array of symbols rather than strings.. but if strings are actually what's being used everywhere, it's not a best practice anymore.

bkiahstroud commented:
@afred do you have changes locally that haven't been pushed up to this PR yet? For example, I don't see a new method named `AdminData#validate_undeletable_fields` that you referenced in [this comment](https://github.com/WGBH-MLA/ams/pull/889#issuecomment-2165869946)

afred committed:
adds missing code that was supposed to be part of the last commit :facepalm:, please squash when merging and remove this msg.

bkiahstroud commented on app/transactions/ams/steps/create_aapb_admin_data.rb:
```diff
@@ -51,22 +51,12 @@ def save_aapb_admin_data(change_set)
-          elsif change_set.fields[k_string].present?
-            AdminData::SERIALIZED_FIELDS.include?(k) ? admin_data.send("#{k}=", Array(change_set.fields[k_string])) : admin_data.send("#{k}=", change_set.fields[k_string].to_s)
-          end
+        admin_data_values = change_set.fields.slice(*AdminData.attributes_for_update).compact
```
`#slice` only works if the keys are the same type. We know `AdminData.attributes_for_update` are strings, is `change_set.fields`?

afred replied:
Yeah, pretty sure from debugging the specs. Also the previous comparison assumed string keys: `change_set.fields[k_string].present?`

bkiahstroud commented on app/transactions/ams/steps/create_aapb_admin_data.rb:
```diff
@@ -51,22 +51,12 @@ def save_aapb_admin_data(change_set)
-
-        # The presence of the key with a "blank" value indicates we're intentionally emptying the value
-        change_set.fields.key?(key) && change_set.fields[key].blank?
+        admin_data.update!(admin_data_values)
```
`#update!` will persist the changes to the database, but I think the purpose of this method is just to set the values. The changes will be persisted by `#save_aapb_admin_data`. I recommend using `#assign_attributes`

afred committed:
Do not persist data in CreateAapbAdminData#set_admin_data_attributes
